File: backend/app/domain/models/dataset.py
from __future__ import annotations

from datetime import UTC, datetime
from enum import Enum
from typing import Any, Dict, List
import uuid

from pydantic import BaseModel, Field


class DatasetStorageType(str, Enum):
    MANAGED_UPLOAD = "managed_upload"
    HOST_PATH = "host_path"


class DatasetFile(BaseModel):
    path: str
    size: int = 0
    role: str = "data"
    content_type: str | None = None

    @property
    def name(self) -> str:
        return self.path


class DatasetLocation(BaseModel):
    location_id: str = Field(default_factory=lambda: f"dsl_{uuid.uuid4().hex[:16]}")
    node_id: str
    storage_type: DatasetStorageType
    source_path: str
    mount_name: str = ""
    read_only: bool = True
    verified: bool = False
    verification_message: str = ""
    version: str = "1"


class DataCenterDataset(BaseModel):
    dataset_id: str = Field(default_factory=lambda: f"ds_{uuid.uuid4().hex[:16]}")
    external_id: str = ""
    data_center_id: str
    data_center_name: str
    name: str
    name_key: str = ""
    description: str = ""
    temporal_coverage: str = ""
    spatial_coverage: str = ""
    data_type: str = ""
    tags: List[str] = Field(default_factory=list)
    preview_url: str = ""
    nc_view_url: str | None = None
    files: List[DatasetFile] = Field(default_factory=list)
    metadata: Dict[str, Any] = Field(default_factory=dict)
    locations: List[DatasetLocation] = Field(default_factory=list)
    enabled: bool = True
    is_submission: bool = False
    created_by: str | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))


class DatasetMount(BaseModel):
    dataset_id: str
    source_id: str = ""
    display_name: str = ""
    node_id: str
    storage_type: DatasetStorageType
    source: str
    target: str
    read_only: bool = True
    version: str = "1"


class MountedDataset(DataCenterDataset):
    sandbox_path: str

File: frontend/src/api/dataset.ts
import { apiClient, type ApiResponse } from './client';

export interface DataCenterDatasetFile {
  name: string;
  path: string;
  size: number;
  role: string;
  content_type?: string | null;
}

export interface DatasetLocation {
  location_id: string;
  node_id: string;
  storage_type: 'managed_upload' | 'host_path';
  read_only: boolean;
  verified: boolean;
  verification_message: string;
  version: string;
}

export interface DataCenterDataset {
  dataset_id: string;
  external_id: string;
  data_center_id: string;
  data_center_name: string;
  name: string;
  description: string;
  temporal_coverage: string;
  spatial_coverage: string;
  data_type: string;
  tags: string[];
  preview_url: string;
  ncViewUrl?: string | null;
  files: DataCenterDatasetFile[];
  metadata: Record<string, unknown>;
  locations: DatasetLocation[];
  enabled: boolean;
  created_by?: string | null;
  created_at: string;
  updated_at: string;
}

export interface DataProductFile {
  file_id: string;
  filename: string;
  relative_path: string;
  role: 'data' | 'chart' | 'source' | 'report' | 'other';
  content_type?: string | null;
  size: number;
  source_artifact_id?: string | null;
  source_tool?: string | null;
  is_primary: boolean;
  created_at?: string | null;
}

export interface DataProduct {
  product_id: string;
  dataset_id: string;
  source_session_id: string;
  name: string;
  description: string;
  generation_method: string;
  created_by: string;
  owner_id?: string | null;
  version: number;
  files: DataProductFile[];
  directories: string[];
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface DatasetSuggestedQuestionsResponse {
  questions: string[];
}

export interface DatasetChatSession {
  session_id: string;
  title: string | null;
  latest_message: string | null;
  latest_message_at: number | null;
  status: 'pending' | 'running' | 'waiting' | 'completed';
}

export async function listDataCenterDatasets(): Promise<DataCenterDataset[]> {
  const response = await apiClient.get<ApiResponse<{ datasets: DataCenterDataset[] }>>('/datasets');
  return response.data.data.datasets;
}

export async function getDataCenterDataset(datasetId: string): Promise<DataCenterDataset> {
  const response = await apiClient.get<ApiResponse<DataCenterDataset>>(
    `/datasets/${encodeURIComponent(datasetId)}`,
  );
  return response.data.data;
}

export async function generateDatasetSuggestedQuestions(datasetId: string): Promise<string[]> {
  const response = await apiClient.post<ApiResponse<DatasetSuggestedQuestionsResponse>>(
    `/datasets/${encodeURIComponent(datasetId)}/suggested-questions`,
  );
  return response.data.data.questions;
}

export async function listDatasetChatSessions(datasetId: string): Promise<DatasetChatSession[]> {
  const response = await apiClient.get<ApiResponse<{ sessions: DatasetChatSession[] }>>(
    `/datasets/${encodeURIComponent(datasetId)}/sessions`,
  );
  return response.data.data.sessions;
}

export async function listDatasetDataProducts(datasetId: string): Promise<DataProduct[]> {
  const response = await apiClient.get<ApiResponse<DataProduct[]>>(
    '/datasets/' + encodeURIComponent(datasetId) + '/data-products',
  );
  return response.data.data;
}

export async function updateDatasetDataProduct(datasetId: string, productId: string, payload: { name: string; description: string; generation_method: string; created_by: string; directories: string[]; files: DataProductFile[] }): Promise<DataProduct> {
  const response = await apiClient.put<ApiResponse<DataProduct>>(
    '/datasets/' + encodeURIComponent(datasetId) + '/data-products/' + encodeURIComponent(productId),
    payload,
  );
  return response.data.data;
}

export async function deleteDatasetDataProduct(datasetId: string, productId: string): Promise<void> {
  await apiClient.delete<ApiResponse<{ deleted: boolean }>>(
    '/datasets/' + encodeURIComponent(datasetId) + '/data-products/' + encodeURIComponent(productId),
  );
}

export async function downloadDatasetDataProduct(datasetId: string, productId: string): Promise<Blob> {
  const response = await apiClient.get(
    '/datasets/' + encodeURIComponent(datasetId) + '/data-products/' + encodeURIComponent(productId) + '/download',
    { responseType: 'blob' },
  );
  return response.data as Blob;
}
